src/audit/routes/query.py

```diff
@@ -0,0 +1,276 @@
+    skipping entries when getting the next page.
+
+    Filters can be added as query strings. Accepted filters include all fields
+    for the queried category, as well as the following special filters:
+    - "groupby" to get counts
+    - "count" to get the number of rows instead of a list
+    - "start" to specify a starting timestamp (inclusive). Default: none
+    - "stop" to specify an end timestamp (exclusive). Default: none
+
+    If queries are time-boxed (depends on the configuration),
+    ("stop" - "start") must be lower than the configured maximum.
+
+    Without filters, this endpoint will return all data within the time-box.
+    Add filters as query strings like this:
+
+        GET /log/presigned_url?a=1&b=2
+
+    This will match all records that have values containing all of:
+
+        {"a": 1, "b": 2}
+
+    Providing the same key with more than one value filters records whose
+    value of the given key matches any of the given values. But values of
+    different keys must all match. For example:
+
+        GET /log/presigned_url?a=1&a2&b=3
+
+    Matches these:
+
+        {"a": 1, "b": 3}
+        {"a": 2, "b": 3}
+
+    But won't match these:
+
+        {"a": 1, "b": 10}
+        {"a": 10, "b": 3}
+
+    `groupby` example:
+
+        GET /log/presigned_url?a=1&groupby=b&groupby=c
+
+        {"b": 1, "c": 2, "count": 5}
+        {"b": 1, "c": 3, "count": 8}
+
+    `count` example:
+
+        GET /log/presigned_url?a=1&groupby=b&groupby=c&count
+
+        Returns: 2 (see previous example returning 2 rows)
+    """
+    logger.debug(f"Querying category {category}")
+
+    if category not in CATEGORY_TO_MODEL_CLASS:
+        raise HTTPException(
+            HTTP_400_BAD_REQUEST,
+            f"Category '{category}' is not one of {list(CATEGORY_TO_MODEL_CLASS.keys())}",
+        )
+    model = CATEGORY_TO_MODEL_CLASS[category]
+
+    # TODO if the category is `presigned_url`, we could implement more granular authz using the audit logs' `resource_paths`.
+    resource_path = f"/services/audit/{category}"
+    await auth.authorize("read", [resource_path])
+
+    # don't just overwrite `stop` with the current timestamp, because the
+    # `stop` param is exclusive and when we don't specify it, we want to
+    # be able to query logs we just created
+    effective_stop = stop or int(time.time())
+
+    timebox_max_seconds = None
+    if config["QUERY_TIMEBOX_MAX_DAYS"]:
+        timebox_max_seconds = config["QUERY_TIMEBOX_MAX_DAYS"] * 86400
+
+        # if the query is time-boxed and `stop` was not specified,
+        # set `stop` to the newest allowed timestamp
+        if start and not stop:
+            stop = start + timebox_max_seconds
+            effective_stop = stop
+
+        # if the query is time-boxed and `start` was not specified,
+        # set `start` to the oldest allowed timestamp
+        if not start:
+            start = max(effective_stop - timebox_max_seconds, 0)
```

Comment:
Not really necessary, but maybe it would help testability/readability in the future to do something like...

```
start, stop, effective_stop = normalize_times_or_whatever(start, stop)
```

Maybe the same below for deriving the query object, but no need to change anything.